[java-inspections] IDEA-300181  Calling stream several times

pyltsin opened the pull request:
I tried to add a new inspection

https://youtrack.jetbrains.com/issue/IDEA-300181/

I hope I did it correctly,
If you find any mistakes I will be happy to fix them.

amaembo commented on java/java-tests/testData/inspection/dataFlow/fixture/ConsumedStream.java:
```diff
@@ -0,0 +1,191 @@
+  public static void test3() {
+    Stream<String> stream = Stream.of("x");
+    for (int i = 0; i < 10; i++) {
+      <warning descr="Stream has already been linked or consumed">stream</warning>.forEach(System.out::println);
```
This message sounds strange, as on the first iteration it's not true. I think we should separate cases when we have 100% problem and we might have a problem. There should be different messages and "might" problems should be suppressed if REPORT_UNSOUND_WARNINGS is false. You can see how it's done e.g. in `reportFailingCasts`.

pyltsin replied:
Fixed. Thank you!

amaembo commented on java/java-tests/testData/inspection/dataFlow/fixture/ConsumedStream.java:
```diff
@@ -0,0 +1,191 @@
+    Stream<String> stream = Stream.of("x");
+    for (int i = 0; i < 10; i++) {
+      if (i == 5) {
+        stream.forEach(System.out::println);
```
I wonder how this works... Do we really know that `i == 5` will happen only once in the whole loop? I don't remember that our analysis supports this. When if we change the condition to `if (i == 5 || i == 7)`?

pyltsin replied:
Loops are not supported exactly. The case `if (i == 5 || i == 7)` doesn't show warning too. It happens, because the interpreter tries to squash 2 state (1 - when `i == 5`, 2 - when `i != 5`) and `stream.linkedOrConsumed` becomes `BOOLEAN` (or `UNKNOWN` now).
This behavior could be changed, but I've decided to save it, because it can reduce false-positive cases. If it is required, I will fix it

amaembo replied:
Well, `i == 5` test was suggested in case you decide to implement this as a separate inspection (matching PSI), though it would work worse in other cases. We may try to fix this later somehow in general, making dataflow analysis to know that in counting loops values don't repeat. This is definitely out of the scope of this test assignment.

Still probably it's good to report this problem (falsely) for now, as the fact of squashing is the implementation detail of the analysis and should not change the warnings. Well, it's up to you, for now you can leave it as is.

amaembo commented on java/java-analysis-impl/src/com/intellij/codeInspection/dataFlow/CustomMethodHandlers.java:
```diff
@@ -193,7 +193,14 @@ private static CustomMethodHandler toValue(DfTypeCustomMethodHandler handler) {
-              toValue((arguments, state, factory, method) -> COLLECTION_SIZE.asDfType(intValue(0)).meet(LOCAL_OBJECT)));
+              toValue((arguments, state, factory, method) -> COLLECTION_SIZE.asDfType(intValue(0)).meet(LOCAL_OBJECT)))
+    //for propagation CONSUMED_STREAM
+    .register(ConsumedStreamUtils.getSkipStreamMatchers(), (args, memState, factory, method) -> args.myQualifier)
```
Probably providing a `-> this` contract would produce the same result? Though it's not exactly specified that it will be so...

pyltsin replied:
Moved to HardcodedContract. Honestly, I don't want to work with External annotation, because as you wrote, it is not specified

amaembo reviewed (changes requested):
Very good in general. Please address the issues in comments. I have some other thoughts:

1. Instead of using boolean, we can create a special data type like DfStreamStateType (probably an enum) with OPEN/UNKNOWN/CONSUMED values. This will allow us to override DfType::correctTypeOnFlush, so that it can change monotonically only (in particular, if CONSUMED state is reached, flushing keeps it CONSUMED). This is similar to what you suggest with `isStable`. Probably we can later add CLOSED state to distinguish CONSUMED and CLOSED and propagate this analysis to other AutoCloseable's. You may experiement with this if you like. I don't know whether this approach will be a fruitful.
2. Assuming that any unknown method returning stream returns it non-consumed and local sounds a good idea. I would not enumerate all the sources in CustomMethodHandler but support any method somehow instead.
3. Returning a consumed stream from the method should indeed generate a warning. Probably, it can be done inside ControlFlowAnalyzer, adding ENSURE instruction before return if returned expression is a stream.

pyltsin commented on java/java-analysis-impl/src/com/intellij/codeInspection/dataFlow/jvm/SpecialField.java:
```diff
@@ -276,6 +276,25 @@ boolean isMyQualifierType(DfType type) {
+
+    @Override
+    public @NotNull DfType getDefaultValue() {
+      return DfType.TOP;
```
Now UNKNOWN is used. Thank you!

pyltsin commented on java/java-analysis-impl/src/com/intellij/codeInspection/dataFlow/SideEffectHandlers.java:
```diff
@@ -133,14 +136,21 @@ private static void collectionReduce(DfaValueFactory factory, DfaMemoryState sta
+  }
+
+  private static void streamConsume(DfaValueFactory factory, DfaMemoryState state, DfaCallArguments arguments) {
+    DfaVariableValue size = tryCast(CONSUMED_STREAM.createValue(factory, arguments.myQualifier), DfaVariableValue.class);
```
Changed to `consumedStream`

pyltsin commented on java/java-analysis-impl/src/com/intellij/codeInspection/dataFlow/CustomMethodHandlers.java:
```diff
@@ -193,7 +193,14 @@ private static CustomMethodHandler toValue(DfTypeCustomMethodHandler handler) {
+    .register(ConsumedStreamUtils.getSkipStreamMatchers(), (args, memState, factory, method) -> args.myQualifier)
+    //many of them are not fully local objects, but they don't affect on CONSUMED_STREAM
+    .register(ConsumedStreamUtils.getStreamGenerators(),
+              toValue((arguments, state, factory, method) -> CONSUMED_STREAM.asDfType(FALSE)
```
Changed. Thank you!

pyltsin commented on java/java-analysis-impl/src/com/intellij/codeInspection/dataFlow/ConsumedStreamUtils.java:
```diff
@@ -0,0 +1,78 @@
+    instanceCall(JAVA_UTIL_STREAM_DOUBLE_STREAM, "filter", "map", "mapToObj", "mapToInt", "mapToLong", "flatMap", "distinct",
+                 "sorted", "peek", "limit", "skip", "takeWhile", "dropWhile", "boxed", "mapMulti"));
+
+  private static final CallMatcher ON_CLOSE_STREAM_MATCHERS = instanceCall(JAVA_UTIL_STREAM_BASE_STREAM, "onClose");
```
Renamed

pyltsin commented on java/java-analysis-impl/src/com/intellij/codeInspection/dataFlow/ConsumedStreamUtils.java:
```diff
@@ -0,0 +1,78 @@
+public final class ConsumedStreamUtils {
+
+  public static final CallMatcher INTERMEDIATE_OPERATION_MATCHERS = anyOf(
+    instanceCall(JAVA_UTIL_STREAM_STREAM, "filter", "map", "mapToInt", "mapToLong", "mapToDouble", "flatMap", "flatMapToInt",
```
Thank you! all cases are joined into one matcher

pyltsin commented:
Thank you for your review.
- I added DfStreamStateType (UNKNOWN, OPEN, CONSUMED) and changed a little logic because of that
- CustomMethodHandler is changed to accept all methods returning stream
- Add ENSURE to check return value. Also few tests are added

amaembo reviewed (changes requested):
Thank you for the update! I added a few more minor comments. Also, GitHub complains that conflict appeared in this PR. Could you please rebase it and solve the conflict? Thanks.

pyltsin commented on java/java-analysis-impl/src/com/intellij/codeInspection/dataFlow/DataFlowInspectionBase.java:
```diff
@@ -69,8 +69,10 @@ public JComponent createOptionsPanel() {
-    node.addContent(new Element("option").setAttribute("name", "SUGGEST_NULLABLE_ANNOTATIONS").setAttribute("value", String.valueOf(SUGGEST_NULLABLE_ANNOTATIONS)));
-    node.addContent(new Element("option").setAttribute("name", "DONT_REPORT_TRUE_ASSERT_STATEMENTS").setAttribute("value", String.valueOf(DONT_REPORT_TRUE_ASSERT_STATEMENTS)));
+    node.addContent(new Element("option").setAttribute("name", "SUGGEST_NULLABLE_ANNOTATIONS")
+                      .setAttribute("value", String.valueOf(SUGGEST_NULLABLE_ANNOTATIONS)));
```
Fixed, thank you

pyltsin commented on java/java-analysis-impl/src/com/intellij/codeInspection/dataFlow/java/ControlFlowAnalyzer.java:
```diff
@@ -842,6 +844,12 @@ else if (closure instanceof PsiLambdaExpression) {
+        if (InheritanceUtil.isInheritor(returnValue.getType(), JAVA_UTIL_STREAM_BASE_STREAM)) {
+          addInstruction(new DupInstruction());
+          addInstruction(new UnwrapDerivedVariableInstruction(SpecialField.CONSUMED_STREAM));
+          addInstruction(new EnsureInstruction(new ConsumedStreamProblem(returnValue), RelationType.NE, DfStreamStateType.CONSUMED, null));
```
I'm not sure, that it is reasonable for this point, because it is about return statement. But I added this exception to StreamChainInliner and changed com.intellij.codeInspection.dataFlow.java.ControlFlowAnalyzer#finishCall to support this. 

New tests are added (I simplified a bit them):
```java
  private static void test12E(Stream<Object> stream) {
    stream.forEach(System.out::println);
    String text = "noError";
    try {
      <warning descr="Stream has already been linked or consumed">stream</warning>.forEach(System.out::println);
    } catch (IllegalStateException e) {
      text = "IllegalStateException";
    } catch (RuntimeException e) {
      text = "RuntimeException";
    }
    if (<warning descr="Condition 'text.equals(\"IllegalStateException\")' is always 'true'">text.equals("IllegalStateException")</warning>) {
      System.out.println("error");
    }
  }
```

pyltsin commented on java/java-analysis-impl/src/com/intellij/codeInspection/dataFlow/java/inst/MethodCallInstruction.java:
```diff
@@ -379,10 +390,20 @@ else if (myContext instanceof PsiMethodReferenceExpression) {
     PsiType type = getResultType();
 
+    if (InheritanceUtil.isInheritor(type, JAVA_UTIL_STREAM_BASE_STREAM)) {
+      DfType customObject = customObject(TypeConstraints.TOP, DfaNullability.fromNullability(myReturnNullability),
```
Unfortunately, `CONSUMED_STREAM.asDfType` created only NOT_NULL type, and I can't find how I can change it with `meet` calls. I want to support also Nullable type, because I want to support this case:
```java

  public static void test11E() {
    Stream<Integer> integerStream = test11Em();
    Stream<Integer> integerStream1 = integerStream.<warning descr="Method invocation 'filter' may produce 'NullPointerException'">filter</warning>(t -> true);
    Stream<Integer> integerStream2 = <warning descr="Stream has already been linked or consumed">integerStream</warning>.filter(t -> true);
  }

  @Nullable
  public static Stream<Integer> test11Em() {
    return null;
  }
```

amaembo replied:
I see. Well, there's DfReferenceType::dropNullability but this indeed complicates things. Ok, let's keep it as is for now, thanks.

pyltsin commented on java/java-analysis-impl/src/com/intellij/codeInspection/dataFlow/HardcodedContracts.java:
```diff
@@ -69,6 +73,20 @@ public static boolean isKnownNoParameterLeak(@Nullable PsiMethod method) {
+   * @param method   method to test
+   * @return true if given method doesn't spoil its qualifier
+   */
+  public static boolean isKnownNoQualifierLeak(@Nullable  PsiMethod method) {
```
Fixed, thank you

pyltsin commented on java/java-analysis-impl/src/com/intellij/codeInspection/dataFlow/HardcodedContracts.java:
```diff
@@ -52,10 +54,12 @@ public final class HardcodedContracts {
    * @return true if given method doesn't spoil the arguments' locality
    */
-  public static boolean isKnownNoParameterLeak(@Nullable PsiMethod method) {
+  public static boolean isKnownNoParameterLeak(@Nullable PsiMethod method, DfaValue value, DfaMemoryState memState) {
```
Fixed, thank you!

pyltsin commented on java/java-analysis-impl/src/com/intellij/codeInspection/dataFlow/types/DfStreamStateType.java:
```diff
@@ -0,0 +1,79 @@
+
+  @Override
+  public @NotNull DfType meetRelation(@NotNull RelationType relationType, @NotNull DfType other) {
+    if (other instanceof DfStreamStateType) {
```
I changed `tryNegate` and `fromRelation(@NotNull RelationType relationType)` to support without overloading meetRelation(

amaembo replied:
Thanks, looks much better now.

pyltsin commented:
Additionally, I changed behavior for squashing to warn these cases.

Thank your for your comments!

amaembo commented:
Merged manually, conflicts resolved. Thanks!
